posix_sockets: port to sock

miri64 opened the pull request:
Ports `posix_sockets` to sock.

emmanuelsearch commented on sys/posix/sockets/posix_sockets.c:
```diff
@@ -170,110 +206,73 @@ static inline int _get_data_from_sockaddr(const struct sockaddr *address, size_t
+            out->addr.ipv4 = in_addr->sin_addr.s_addr;
+            out->port = ntohs(in_addr->sin_port);
             break;
+#ifdef SOCK_HAS_IPV6
```
Any reason while you ifdef like this here?

miri64 replied:
Because without `SOCK_HAS_IPV6` defined `sock` does not support IPv6 so without this `#ifdef` the code below would be invalid.

emmanuelsearch replied:
I understand #ifdef works ;) but I was wondering because it makes IPv4 look mandatory and IPv6 look optional, whereas in the networks we target, it's usually the reverse... hence my question.

miri64 replied:
Well in sock it is that way, because it is not just build for networks we target but also as a simplification for POSIX sockets (see [kaspar030/sock](https://github.com/kaspar030/sock/tree/master/src/posix)). And from an application point of view it does make sense, since IPv4 is always integrated in some capacity in IPv6 while the reverse is not the case + IPv6 needs more memory (due to its long addresses) so optimizing it out if necessary is good.

emmanuelsearch commented:
Quickly tested on SAMR21, one running the POSIX example and on running the GNRC_networking example.

Works as expected for the sending (from POSIX example to GNRC_networking example). 

But for receiving (from GNRC_networking example to POSIX example) it just outputs the packet dump, not the udp server output. Is this intended?

udp send seems to crash 50% of the time (in both examples mentioned above).

It seems there's a significant number of UDP content that does not arrive (not only to send crash).

Also tested POSIX example to POSIX example, did not work much better.

Can you test again on your end?

miri64 commented:
> But for receiving (from GNRC_networking example to POSIX example) it just outputs the packet dump, not the udp server output. Is this intended?

What dou you mean by udp server output? It should just dump the message you send it.

miri64 commented:
> Can you test again on your end?

Will do, but not right now.

emmanuelsearch commented:
More precisely udp send is causing kernel panic 50% of the time.
Could be in the cases where sending while a UDP server is concurrently running  (fuzzy guess).

For the output, I get sometimes:

```
2016-11-02 12:37:37,814 - INFO # > udp server start 1337
2016-11-02 12:37:37,817 - INFO # Success: started UDP server on port 1337
2016-11-02 12:37:50,666 - INFO # > Received data: hey
```

Some other (rare) times I just get a dump of packet snips.

PeterKietzmann commented:
Compiling examples/posix_sockets fails
```
posix_sockets.c:179:38: error: assignment makes integer from pointer without a cast [-Werror]
             in_addr->sin_addr.s_addr = ep->addr.ipv4;
                                      ^
posix_sockets.c:210:28: error: assignment to expression with array type
             out->addr.ipv4 = in_addr->sin_addr.s_addr;
```

PeterKietzmann commented on sys/posix/sockets/posix_sockets.c:
```diff
@@ -737,217 +765,139 @@ ssize_t recvfrom(int socket, void *restrict buffer, size_t length, int flags,
-                errno = -res;
-                return -1;
-            }
+            res = sock_ip_send(&s->sock->raw, buffer, length,
```
Missing protocol parameter here

miri64 replied:
Fixed

miri64 commented:
Rebased and addressed comments

@emmanuelsearch 

> Some other (rare) times I just get a dump of packet snips.

That's weird... `gnrc_pktdump` isn't even compiled into `examples/posix_sockets` oO.

miri64 commented:
Btw with this #5802 and #6012 we could get rid of `conn` (though I would propose to do that in any case for the 2017.04 release at the latest).

emmanuelsearch commented:
Tried to build for native on my Linux VM and on OS X. Both builds fail. 

Here's the error I get on Linux:
```
/home/vagrant/RIOT/examples/posix_sockets/bin/native/cpu.a(irq_cpu.o): In function `native_isr_entry':
irq_cpu.c:(.text.native_isr_entry+0x12f): undefined reference to `_native_sig_leave_tramp'
/home/vagrant/RIOT/examples/posix_sockets/bin/native/cpu.a(native_cpu.o): In function `_native_mod_ctx_leave_sigh':
native_cpu.c:(.text._native_mod_ctx_leave_sigh+0x23): undefined reference to `_native_sig_leave_handler'
collect2: error: ld returned 1 exit status
/home/vagrant/RIOT/examples/posix_sockets/../../Makefile.include:271: recipe for target 'all' failed
```

and here the errors & warnings I get on OS X

```
/Users/emmanuel/Documents/work/CODING/RIOT/sys/posix/sockets/posix_sockets.c:245:18: error: variable 'res' is used uninitialized
      whenever switch case is taken [-Werror,-Wsometimes-uninitialized]
            case SOCK_DGRAM:
                 ^~~~~~~~~~
/Users/emmanuel/Documents/work/CODING/RIOT/sys/posix/include/sys/socket.h:89:25: note: expanded from macro 'SOCK_DGRAM'
#define SOCK_DGRAM      (1)     /**< Datagram socket */
                        ^~~
/Users/emmanuel/Documents/work/CODING/RIOT/sys/posix/sockets/posix_sockets.c:276:12: note: uninitialized use occurs here
    return res;
           ^~~
/Users/emmanuel/Documents/work/CODING/RIOT/sys/posix/sockets/posix_sockets.c:241:9: error: variable 'res' is used uninitialized
      whenever 'if' condition is false [-Werror,-Wsometimes-uninitialized]
    if (s->sock != NULL) {
        ^~~~~~~~~~~~~~~
/Users/emmanuel/Documents/work/CODING/RIOT/sys/posix/sockets/posix_sockets.c:276:12: note: uninitialized use occurs here
    return res;
           ^~~
/Users/emmanuel/Documents/work/CODING/RIOT/sys/posix/sockets/posix_sockets.c:241:5: note: remove the 'if' if its condition is always
      true
    if (s->sock != NULL) {
    ^~~~~~~~~~~~~~~~~~~~~
/Users/emmanuel/Documents/work/CODING/RIOT/sys/posix/sockets/posix_sockets.c:235:12: note: initialize the variable 'res' to silence
      this warning
    int res;
           ^
            = 0
2 errors generated.
```

emmanuelsearch commented:
Builds for the SAMR21, though.

miri64 commented:
Looks more like local native foobar. Are you at the office tomorrow so I can take a look?

miri64 commented:
(Have the same issue on milhouse though, will look into it).

miri64 commented:
https://github.com/RIOT-OS/RIOT/pull/6004/commits/c0b9c70f5298d3c3a43541026ff41c5530ebf8c7 should fix the OSX issue.

emmanuelsearch commented:
Tested on SAMR21. Works.

miri64 commented:
Fixed test applications depending on POSIX sockets.

miri64 commented:
And rebased.

emmanuelsearch commented:
retested. Works on native and on SAMR21 for a single ping.
However, when I specify say, 10 pings 10ms appart, it does not display the output in "real time", I have to somehow type another command on the sender side before anything happens in the terminal. 
Is that intended?

miri64 commented:
Will look into that, but might be unrelated to the actual port (and already have existed before)

miri64 commented:
@emmanuelsearch could not reproduce

```
2017-01-24 14:09:29,564 - INFO #  Received data: Hallo Welt
2017-01-24 14:09:48,414 - INFO # Received data: Hallo Welt
2017-01-24 14:09:48,432 - INFO # Received data: Hallo Welt
2017-01-24 14:09:48,446 - INFO # Received data: Hallo Welt
```

(First one was send singleton, the 3 remaining in batch)

miri64 commented:
Addressed issues pointed out by `cppcheck`.

emmanuelsearch commented:
could not reproduce again... Tried to hammer, seems to held well.

miri64 commented:
Squashed

miri64 commented:
removed duplicate introduction of sock pseudomodules and squashed immediately

PeterKietzmann commented:
All green!
